runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.beam.sdk.transforms.reflect.DoFnSignature;
+import org.apache.beam.sdk.transforms.reflect.DoFnSignatures;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.WindowTracing;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.util.state.State;
+import org.apache.beam.sdk.util.state.StateSpec;
+import org.joda.time.Instant;
+
+/**
+ * A customized {@link DoFnRunner} that handles late data dropping and
+ * garbage collect for stateParDo. It registers a GC timer in
+ * processElement and does cleanup in onTimer.
+ *
+ * @param <InputT> the type of the {@link DoFn} (main) input elements
+ * @param <OutputT> the type of the {@link DoFn} (main) output elements
+ */
+public class StatefulDoFnRunner<InputT, OutputT> implements DoFnRunner<InputT, OutputT> {
+
+  public static final String GC_TIMER_ID = "StatefulParDoGcTimerId";
+  public static final String DROPPED_DUE_TO_LATENESS_COUNTER = "StatefulParDoDropped";
+
+  private DoFn<InputT, OutputT> fn;
+  private final SimpleDoFnRunner<InputT, OutputT> doFnRunner;
+  private final StepContext stepContext;
+  private final WindowingStrategy<?, ?> windowingStrategy;
+  private final Aggregator<Long, Long> droppedDueToLateness;
+  private final Coder<BoundedWindow> windowCoder;
+  private final DoFnSignature signature;
+
+  @SuppressWarnings("unchecked")
+  public StatefulDoFnRunner(
+      DoFn<InputT, OutputT> fn,
+      SimpleDoFnRunner<InputT, OutputT> doFnRunner,
+      StepContext stepContext,
+      AggregatorFactory aggregatorFactory,
+      WindowingStrategy<?, ?> windowingStrategy) {
+    this.fn = fn;
+    this.doFnRunner = doFnRunner;
+    this.stepContext = stepContext;
+    this.windowingStrategy = windowingStrategy;
+    this.signature = DoFnSignatures.getSignature(fn.getClass());
+    windowCoder = (Coder<BoundedWindow>) windowingStrategy.getWindowFn().windowCoder();
+    this.droppedDueToLateness = aggregatorFactory.createAggregatorForDoFn(
+        fn.getClass(), stepContext, DROPPED_DUE_TO_LATENESS_COUNTER,
+        Sum.ofLongs());
+  }
+
+  @Override
+  public void startBundle() {
+    doFnRunner.startBundle();
+  }
+
+  @Override
+  public void processElement(WindowedValue<InputT> compressedElem) {
+
+    TimerInternals timerInternals = stepContext.timerInternals();
+
+    // StatefulDoFnRunner is always observes window
+    for (WindowedValue<InputT> value : compressedElem.explodeWindows()) {
+
+      BoundedWindow window = value.getWindows().iterator().next();
+      Instant gcTime = window.maxTimestamp().plus(windowingStrategy.getAllowedLateness());
+
+      if (!dropLateData(window)) {
+        timerInternals.setTimer(StateNamespaces.window(windowCoder, window),
+            GC_TIMER_ID, gcTime, TimeDomain.EVENT_TIME);
+        doFnRunner.invokeProcessElement(value);
```

Comment:
I think we should invoke `processElement()`. This does potentially more work because it will try to explode the one window again but we never know what other work might be going on in `processElement()` of the wrapped `DoFnRunner`.